### PR TITLE
Add PythonQuandl

### PR DESCRIPTION
**Pull request by Marinovsky**
- Add `PythonQuandl.cs` to keep backwards compatibility with LEAN
- Fix bug in `NasdaqDataLink.cs`. `Intersect()` method keeps the order of the current object is calling the method.
- Fix bug in `ValueColumn()` test. `Reader()` method returns an instance with the data is reading

### CODE REVIEW

File: QuantConnect.DataSource.csproj [hunk 1]
```diff
@@ -3,11 +3,12 @@
   <PropertyGroup>
     <TargetFramework>net5.0</TargetFramework>
     <RootNamespace>QuantConnect.DataSource</RootNamespace>
-    <DocumentationFile>$(OutputPath)\QuantConnect.DataSource.xml</DocumentationFile>
+    <RootNamespace>QuantConnect.DataSource.NasdaqDataLink</RootNamespace>
+    <DocumentationFile>C:\Users\JUAN\source\repos\Lean.DataSource.NasdaqDataLink\QuantConnect.DataSource.xml</DocumentationFile>
```
Comment (Martin-Molinero) [inline QuantConnect.DataSource.csproj]: >  <DocumentationFile>C:\Users\JUAN\source\repos\Lean.DataSource.NasdaqDataLink\QuantConnect.DataSource.xml</DocumentationFile>

File: PythonNasdaq.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+*/
+
+namespace QuantConnect.DataSource
+{
+    /// <summary>
+    /// Dynamic data class for Python algorithms.
+    /// </summary>
+    public class PythonNasdaq : NasdaqDataLink
+    {
+        protected static string ValueColumnName = "Close";
```
Comment (Martin-Molinero) [inline PythonNasdaq.cs]: hm this can't be static what if we want 2 different Nasdaq sources, why is this required?

File: NasdaqImporterAlgorithm.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+            self.Debug("Purchased {0} >> {1}".format(self.nasdaqCode, self.Time))
+
+        self.Plot(self.nasdaqCode, "PriceSMA", self.sma.Current.Value)
+
+# NasdaqDataLink often doesn't use close columns so need to tell LEAN which is the "value" column.
+class NasdaqkCustomColumns(PythonNasdaq):
+    '''Custom nasdaq data type for setting customized value column name. Value column is used for the primary trading calculations and charting.'''
+    def __init__(self):
+        # Define ValueColumnName: cannot be None, Empty or non-existant column name
+        self.ValueColumnName = "adj. close"
```
Comment (Martin-Molinero) [inline NasdaqImporterAlgorithm.py:51]: > "adj. close"

Just checking, is the space here right?

File: MortgageRateVolatilityAlgorithm.cs [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+                });
+            }
+        }
+        public class NasdaqMortgagePriceColumns : NasdaqDataLink
+        {
+            public NasdaqMortgagePriceColumns()
+
+                // Rename the NasdaqDataLink object column to the data we want, which is the 'Value' column
+                // of the CSV that our API call returns
+                : base(valueColumnName: "Value")
```
Comment (Martin-Molinero) [inline MortgageRateVolatilityAlgorithm.cs]: I wonder if all there classes are actually now required, given we added a default `_keywords` collection?

File: DemostrationWithIndicator.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        # Wait for all indicators to fully initialize
+        if not (self.ibm_sma.IsReady and self.fb_sma.IsReady and self.ratio.IsReady): return
+        if not self.Portfolio.Invested and self.ratio.Current.Value > 1:
+            self.MarketOrder(self.ibm, 100)
+        elif self.ratio.Current.Value < 1:
+                self.Liquidate()
+
+class NasdaqCustomColumns(PythonNasdaq):
+    def __init__(self):
+        self.ValueColumnName = "adj. close"
```
Comment (Martin-Molinero) [inline DemostrationWithIndicator.py:66]: > "adj. close"

Similar comment about the space

**Review by Martin-Molinero — COMMENTED** [review]
Cool! Leaving minor comments 🎇

File: MortgageRateVolatilityAlgorithm.cs (hunk 4 above)
Comment (Marinovsky) [reply MortgageRateVolatilityAlgorithm.cs]: The problem is that some of the assets have more than one column present in `_keywords` like `SHFE/SCF2021`

File: NasdaqDataLink.cs [hunk 6]
```diff
@@ -32,12 +32,36 @@ namespace QuantConnect.DataSource
...
+
+            if (string.IsNullOrWhiteSpace(_valueColumnName)) return;
+
+            // Insert the value column name at the beginning of the keywords list
+            _keywords.Insert(0, _valueColumnName);
+        }
+
+        /// <summary>
+        /// Name of the column is going to be used for the field Value
+        /// </summary>
```
Comment (Martin-Molinero) [inline NasdaqDataLink.cs:45]: Would be nice to add _why_ is this here

File: NasdaqDataLink.cs [hunk 7]
```diff
@@ -32,12 +32,36 @@ namespace QuantConnect.DataSource
...
         private readonly List<string> _propertyNames = new List<string>();
 
         // The NasdaqDataLink will use one of these column names if they are available and another option is not provided
         private readonly List<string> _keywords = new List<string> { "close", "price", "settle", "value" };
 
+        /// <summary>
+        /// Inserts the name of the column at first position in _keywords list
+        /// </summary>
+        /// <param name="valueColumnName">Name of the column to be used as Value</param>
+        private void SetValueColumnName(string valueColumnName)
```
Comment (Martin-Molinero) [inline NasdaqDataLink.cs]: nit - `private` should go after `public`

File: NasdaqDataLink.cs [hunk 8]
```diff
@@ -51,7 +75,7 @@ static NasdaqDataLink()
         /// <summary>
         /// Default <see cref="NasdaqDataLink"/> constructor uses Close as its value column
         /// </summary>
-        public NasdaqDataLink() : this("Close")
+        public NasdaqDataLink()
```
Comment (Martin-Molinero) [inline NasdaqDataLink.cs:85]: >  : this("Close")

This default value should still be handled I think, maybe calling `SetValueColumnName("Close")` in the body?

File: tests/NasdaqDataLinkTests.cs [hunk 9]
```diff
@@ -95,6 +96,75 @@ public void ValueColumn()
...
+            var data = instance.Reader(config, $"2021-12-02,100,101,100,101,1000,{expected}", DateTime.UtcNow, false);
+
+            Assert.AreEqual(expected, data.Value);
+        }
+		
+		[Test]
+        public void TwoPythonValueColumn()
+        {
+            var ibmExpected = 999;
+            var spyExpected = 111;
```
Comment (Martin-Molinero) [inline tests/NasdaqDataLinkTests.cs:130]: Cool!

File: NasdaqDataLink.cs [hunk 10]
```diff
@@ -57,13 +59,26 @@ static NasdaqDataLink()
             // The NasdaqDataLink API now requires TLS 1.2 for API requests (since 9/18/2018).
             // NET 4.5.2 and below does not enable this more secure protocol by default, so we add it in here
             ServicePointManager.SecurityProtocol |= SecurityProtocolType.Tls12;
+
+            var potentialNasdaqToken = Config.Get("nasdaq-auth-token");
+            var potentialQuandlToken = Config.Get("quandl-auth-token");
+
+            if (potentialNasdaqToken.Length != 0)
```
Comment (Martin-Molinero) [inline NasdaqDataLink.cs]: - Should use `string.IsNullOrEmpty()` instead, else default value for string will be `null` and throw
- Suggest directly moving `Config.Get("quandl-auth-token")` into the `else if` bellow, so that in the happy case it doesn't log anything (config.get will log if no value found)

File: NasdaqDataLink.cs [hunk 11]
```diff
@@ -57,13 +59,26 @@ static NasdaqDataLink()
...
+            {
+                SetAuthCode(potentialNasdaqToken);
+            } 
+            else if (potentialQuandlToken.Length != 0)
+            {
+                SetAuthCode(potentialQuandlToken);
+                Log.Error("Quandl is obsolete. Use NasdaqDataLink instead.");
```
Comment (Martin-Molinero) [inline NasdaqDataLink.cs]: Log error should follow logging style -> `"NasdaqDataLink(): 'quandl-auth-token' is obsolete please use 'nasdaq-auth-token' instead."`
